core/src/main/java/com/graphhopper/routing/lm/LandmarkStorage.java

```diff
@@ -0,0 +1,745 @@
+import com.graphhopper.routing.subnetwork.SubnetworkStorage;
+import com.graphhopper.routing.subnetwork.TarjansSCCAlgorithm;
+import com.graphhopper.routing.util.*;
+import com.graphhopper.routing.util.spatialrules.Polygon;
+import com.graphhopper.routing.util.spatialrules.SpatialRule;
+import com.graphhopper.routing.util.spatialrules.SpatialRuleLookupArray;
+import com.graphhopper.routing.util.spatialrules.countries.DefaultSpatialRule;
+import com.graphhopper.routing.weighting.AbstractWeighting;
+import com.graphhopper.routing.weighting.ShortestWeighting;
+import com.graphhopper.routing.weighting.Weighting;
+import com.graphhopper.storage.*;
+import com.graphhopper.util.*;
+import com.graphhopper.util.exceptions.ConnectionNotFoundException;
+import com.graphhopper.util.shapes.BBox;
+import com.graphhopper.util.shapes.GHPoint;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * This class stores the landmark nodes and the weights from and to all other nodes in every
+ * subnetwork. This data is created to apply a speed-up for path calculation but at the same times
+ * stays flexible to per-request changes. The class is safe for usage from multiple reading threads
+ * across algorithms.
+ *
+ * @author Peter Karich
+ */
+public class LandmarkStorage implements Storable<LandmarkStorage> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LandmarkStorage.class);
+    private static final int UNSET_SUBNETWORK = -1;
+    private static final int UNCLEAR_SUBNETWORK = 0;
+    // one node has an associated landmark information ('one landmark row'): the forward and backward weight
+    private long LM_ROW_LENGTH;
+    private int landmarks;
+    private final int FROM_OFFSET;
+    private final int TO_OFFSET;
+    private final DataAccess landmarkWeightDA;
+    /* every subnetwork has its own landmark mapping but the count of landmarks is always the same */
+    private final List<int[]> landmarkIDs;
+    private double factor = 1;
+    private final static double DOUBLE_MLTPL = 1e6;
+    private final GraphHopperStorage graph;
+    private final FlagEncoder encoder;
+    private final Weighting weighting;
+    private Weighting lmSelectionWeighting;
+    private final TraversalMode traversalMode;
+    private boolean initialized;
+    private int minimumNodes = 500_000;
+    private SubnetworkStorage subnetworkStorage;
+    private Map<String, Polygon> borderMap = new HashMap<>();
+
+    public LandmarkStorage(GraphHopperStorage graph, Directory dir, int landmarks, final Weighting weighting, TraversalMode traversalMode) {
+        this.graph = graph;
+        this.minimumNodes = Math.min(graph.getNodes() / 2, 500_000);
+        this.encoder = weighting.getFlagEncoder();
+        this.weighting = weighting;
+        this.lmSelectionWeighting = new ShortestWeighting(encoder) {
```

Comment:
What do you mean here? The tricky part is that we are using the normal weighting for the preparation but the landmark selection heuristic works best with shortest. See phase 1 vs. 2 in createLandmarks